### DISCUSSION TITLE
Per-hash achievement definitions

### DISCUSSION THREAD

**Discussion by dude1286** [RetroAchievements/RAIntegration#783] [Ideas]
One of the biggest questions is when starting up a game in RA is wondering whether or not the ROM you have will work on the set list that has been developed for. Sometimes, there are no differences when it comes to the RAM when playing a game from different regions, sometimes it is minimal so it is easy to create alts for each region, while others it is impossible due to the amount of differences. This is why I suggest change the data type of the attribute in the JSON file Achievements[{MemAddr}] from a string to a hash, where all the current JSON files MemAddr would have the key Default, while any other requirement sets would be named after the hash of the ROM involved. This would make region differences easier to develop for, and maybe could be used for the ROMs of the GBA Classic NES series to be linked to their NES counterparts.

**Comment by Jamiras** [comment]
See https://github.com/RetroAchievements/RAWeb/issues/717 and https://github.com/RetroAchievements/RAWeb/issues/814, which should provide a quick overview of whether or not a ROM is supported and how compatible it is.

The problem with defining unique MemAddrs per MD5 is keeping them sync'd when fixing tickets. When cheat protection is added to the original set, it would be easy to forget to add it to the secondary copies of the achievements.

Something I've been contemplating for a while is variables. These would be defined externally from the achievements/leaderboards and could just be referenced by them - similar to how the achievements/leaderboards currently reference memory. "8-bit Variable FOO = 6". The thought is that repeated condition chains for complex calculations (mainly AddAddress/AddSource chains) could be calculated once and referenced by multiple achievements. The logic could also be extended to support differences between regions/versions - either through logic, or by allowing individual MD5s to override the logic as proposed above. This way, the achievement "logic" is unaffected by changes made to support regional differences.

**Comment by dude1286** [comment]
Maybe something along the lines of a hash of hashes for the variables. Using the below as an example from yaml format (I am more familiar with yaml, but json files can be loaded properly with a yaml library, so they are very similar).

```
variables:
  default:
    LIVES: '0x00a0'
  fer091245:
    LIVES: '0x00b0'
```

The above is just a quick idea for a variable for the LIVES address for a default and a specific hash. The only issue with going with this idea over the idea that I posted earlier is that maybe the logic is slightly changed due to a bug fix could make the logic much more different that could make the original logic be difficult to rework for the separate hash.

I don't know if C++ has something similar, but in the Ruby computer language, you can perform string % hash, where Ruby will naturally change all instances of the keys in the hash into the corresponding value inside the string. FYI, a Ruby hash is the equivalent of a C++ map, which is pretty much a key, value pairing. If done in Ruby, the above would be something along these lines:
```
json_file['Achievements'].each do |cheevo|
  cheevo['MemAddr'] = cheevo['MemAddr'] % json_file['variables']
end
```
